art/steps.py

```diff
@@ -570,3 +577,111 @@ def do(self, previous_states: Dict):
+
+    def __init__(
+        self,
+        model: ArtModule,
+        model_modifiers: List[Callable] = [],
+        logger: Optional[Union[Logger, Iterable[Logger], bool]] = None,
+        freezed_trainer_kwargs: Dict = {},
+        unfreezed_trainer_kwargs: Dict = {},
+        freeze_names: Optional[list[str]] = None,
+        freeze_to: Optional[int] = None,
+        fine_tune_lr: float = 1e-5,
+        fine_tune: bool = True,
+    ):
+        """
+        This method initializes the step
+
+        Args:
+            model (ArtModule): model
+            model_modifiers (List[Callable], optional): model modifiers. Defaults to [].
+            logger (Optional[Union[Logger, Iterable[Logger], bool]], optional): logger. Defaults to None.
+            freezed_trainer_kwargs (Dict, optional): trainer kwargs use for transfer learning with freezed weights. Defaults to {}.
+            unfreezed_trainer_kwargs (Dict, optional): trainer kwargs use for fine tuning with unfreezed weights. Defaults to {}.
+            freeze_names (Optional[list[str]], optional): name of model to freeze which appears in layers. Defaults to None.
+            freeze_to (Optional[int], optional): number of layers to freeze (from the 0th). Defaults to None.
+            fine_tune_lr (float, optional): fine tune lr. Defaults to 1e-5.
+            fine_tune (bool, optional): whether or not perform fine tuning. Defaults to True.
+        """
+        super().__init__(model, trainer_kwargs=freezed_trainer_kwargs, logger=logger, model_modifiers=model_modifiers)
+        self.freeze_names = freeze_names
+        self.freeze_to = freeze_to
+        self.unfreezed_trainer_kwargs = unfreezed_trainer_kwargs
+        self.fine_tune_lr = fine_tune_lr
+        self.fine_tune = fine_tune
+
+    def do(self, previous_states: Dict):
+        """
+        This method trains the model
+        Args:
+            previous_states (Dict): previous states
+        """
+        self.add_freezing()
+        self.train(trainer_kwargs={"datamodule": self.datamodule})
+        if self.fine_tune:
+            self.add_unfreezing()
+            self.reset_trainer(logger=self.trainer.logger, trainer_kwargs=self.unfreezed_trainer_kwargs)
+            self.train(trainer_kwargs={"datamodule": self.datamodule})
+
+    def log_params(self, model):
+        self.results["parameters"].update(self.trainer_kwargs)
+        super().log_params(model)
+
+    def get_check_stage(self):
+        """Returns check stage"""
+        return TrainingStage.VALIDATION.value
+
+    def add_freezing(self):
+        """Adds freezing to the model"""
+
+        def freeze_by_name(model):
+            """Freeze parameters by layer names."""
+            for name in self.freeze_names:
+                for param in model.named_parameters():
+                    if name in param[0]:
+                        param[1].requires_grad = False
+
+        def freeze_to_layer(model):
```

Comment:
OK, but I think telling it to not freeze last n layers would be more usefull

Reply:
Okay, I may change it.

Reply:
Yes it would be easier for user to tell up to which not to freeze. Now you force user to tell up to which freeze. Probably this is harder.